### PR TITLE
大佬，來抽煙 (Yakuza Smoking Mechanics)

### PR DESCRIPTION
**Pull request by alsoandanswer**
- rscadd: "You can now offer someone a smoke directly from the packet. Just target the mouth area with an open cigarette packet, like you would yourself."
  - rscadd: "You can also offer to light someone else's cigarettes."

### CODE REVIEW

**Comment by alsoandanswer** [comment]
!review

File: code/game/objects/items/weapons/storage/fancy.dm [hunk 1]
```diff
@@ -322,16 +322,29 @@
...
-		to_chat(user, SPAN_NOTICE("The [src] is closed."))
+		to_chat(user, SPAN_NOTICE("\The [src] is closed."))
 		return
 
-	if(M == user && target_zone == BP_MOUTH && contents.len > 0 && !user.wear_mask)
-		var/obj/item/clothing/mask/smokable/cigarette/W = new cigarette_to_spawn(user)
-		if(!istype(W))
-			to_chat(user, SPAN_NOTICE("The [W] is in the way."))
+	if(target_zone == BP_MOUTH && contents.len > 0)
+		var/obj/item/clothing/mask/smokable/cigarette/W = new cigarette_to_spawn(M)
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/storage/fancy.dm:332]: the cigarette shouldn't be made inside of M

File: code/game/objects/items/weapons/storage/fancy.dm [hunk 2]
```diff
@@ -322,16 +322,29 @@
...
+		if(!(M == user))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/storage/fancy.dm:333]:
```suggestion
		if(M != user)
```

File: code/game/objects/items/weapons/storage/fancy.dm [hunk 3]
```diff
@@ -322,16 +322,29 @@
...
+				M.visible_message(SPAN_NOTICE("<b>[M]</b> pushes [user]'s [src] away."))
+				return
+			if(!M.Adjacent(user))
+				to_chat(user, SPAN_WARNING("You need to stay in reaching distance while giving an object."))
+				to_chat(M, SPAN_WARNING("\The [user] moved too far away."))
+				return
+		if(!istype(W) || M.wear_mask)
+			to_chat(user, SPAN_NOTICE("\The [M.wear_mask] is in the way."))
+			if(!(M == user))
+				to_chat(M, SPAN_NOTICE("\The [M.wear_mask] is in the way."))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/storage/fancy.dm:347]: put these two checks earlier

also `!(M == user)` -> `M != user`

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/cigs_lighters.dm [hunk 4]
```diff
@@ -935,10 +935,20 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 		if(M == user)
 			cig.attackby(src, user)
 		else
+			var/response = ""
+			user.visible_message(SPAN_NOTICE("\The <b>[user]</b> holds up \the [src] to \the [M]'s mouth."), SPAN_NOTICE("You hold up \the [src] to \the [M]'s mouth, waiting for them to accept."))
+			response = alert(M, "\The [user] offers to light your [cig.name]. Do you accept?", "Lighter offer", "Accept", "Decline")
```
Comment (SleepyGemmy) [inline code/game/objects/items/weapons/cigs_lighters.dm:940]: can this be spammed?

File: code/game/objects/items/weapons/cigs_lighters.dm [hunk 5]
```diff
@@ -935,10 +935,20 @@ CIGARETTE PACKETS ARE IN FANCY.DM
...
+			if(response != "Accept")
+				M.visible_message(SPAN_NOTICE("<b>[M]</b> pushes [user]'s [src] away."))
+				return
+			if(!M.Adjacent(user))
```
Comment (SleepyGemmy) [inline code/game/objects/items/weapons/cigs_lighters.dm:944]: `use_check_and_message`?

File: code/game/objects/items/weapons/storage/fancy.dm [hunk 6]
```diff
@@ -322,16 +322,29 @@
...
-	if(M == user && target_zone == BP_MOUTH && contents.len > 0 && !user.wear_mask)
-		var/obj/item/clothing/mask/smokable/cigarette/W = new cigarette_to_spawn(user)
-		if(!istype(W))
-			to_chat(user, SPAN_NOTICE("The [W] is in the way."))
+	if(target_zone == BP_MOUTH && contents.len > 0)
+		var/obj/item/clothing/mask/smokable/cigarette/W = new cigarette_to_spawn(M)
+		if(M != user)
+			var/response = ""
+			user.visible_message(SPAN_NOTICE("\The <b>[user]</b> holds up \the [src] to \the [M]'s mouth."), SPAN_NOTICE("You hold up \the [src] to \the [M]'s mouth, waiting for them to accept."))
+			response = alert(M, "\The [user] offers you \a [W.name]. Do you accept?", "Smokable offer", "Accept", "Decline")
```
Comment (SleepyGemmy) [inline code/game/objects/items/weapons/storage/fancy.dm:336]: ditto (comment 1).

File: code/game/objects/items/weapons/storage/fancy.dm [hunk 7]
```diff
@@ -322,16 +322,29 @@
...
+			if(response != "Accept")
+				M.visible_message(SPAN_NOTICE("<b>[M]</b> pushes [user]'s [src] away."))
+				return
+			if(!M.Adjacent(user))
```
Comment (SleepyGemmy) [inline code/game/objects/items/weapons/storage/fancy.dm:340]: ditto (comment 2).

File: code/game/objects/items/weapons/cigs_lighters.dm (hunk 4 above)
Comment (alsoandanswer) [reply code/game/objects/items/weapons/cigs_lighters.dm:940]: works just like the give verb, this should not be a problem

**Review by SleepyGemmy — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Review by FluffyGhoster — APPROVED** [review]

**Pull request merged**